src/app/diamond-posts-page/diamond-posts/diamond-posts.component.ts

```diff
@@ -0,0 +1,166 @@
+        if (startIdx > endIdx) {
+          success([]);
+          return;
+        }
+        const startPage = Math.floor(startIdx / DiamondPostsComponent.PAGE_SIZE);
+        const endPage = Math.floor(endIdx / DiamondPostsComponent.PAGE_SIZE);
+
+        const pageRequests: any[] = [];
+        for (let i = startPage; i <= endPage; i++) {
+          const existingRequest = this.pagedRequests[i];
+          if (existingRequest) {
+            pageRequests.push(existingRequest);
+          } else {
+            const newRequest = this.pagedRequests[i - 1].then((_) => {
+              return this.getPage(i);
+            });
+            this.pagedRequests[i] = newRequest;
+            pageRequests.push(newRequest);
+          }
+        }
+
+        return Promise.all(pageRequests).then((pageResults) => {
+          pageResults = pageResults.reduce((acc, result) => [...acc, ...result], []);
+          const start = startIdx - startPage * DiamondPostsComponent.PAGE_SIZE;
+          const end = start + endIdx - startIdx + 1;
+          return pageResults.slice(start, end);
+        });
+      },
+      settings: {
+        startIndex: 0,
+        minIndex: 0,
+        bufferSize: 10,
+        windowViewport: true,
+      },
+    });
+  }
+
+  getPage(page: number) {
+    if (this.lastPage != null && page > this.lastPage) {
+      return [];
+    }
+    this.loadingNextPage = true;
+    const lastPostHashHex = this.pagedKeys[page];
+    return this.backendApi
+      .GetDiamondedPosts(
+        this.globalVars.localNode,
+        "",
+        this.receiverUsername,
+        "",
+        this.senderUsername,
+        this.globalVars.loggedInUser?.PublicKeyBase58Check,
+        lastPostHashHex,
+        DiamondPostsComponent.PAGE_SIZE
+      )
+      .toPromise()
+      .then((res) => {
+        const posts: PostEntryResponse[] = res.DiamondedPosts;
+        this.pagedKeys[page + 1] = posts.length > 0 ? posts[posts.length - 1].PostHashHex : "";
+        this.lastDiamondLevelOnPage[page] = posts.length > 0 ? posts[posts.length - 1].DiamondsFromSender : -1;
+        if (!posts || posts.length < DiamondPostsComponent.PAGE_SIZE || this.pagedKeys[page + 1] === "") {
+          this.lastPage = page;
+        }
+        if (!this.receiverProfileEntryResponse) {
+          this.receiverProfileEntryResponse = res.ReceiverProfileEntryResponse;
+        }
+        if (!this.senderProfileEntryResponse) {
+          this.senderProfileEntryResponse = res.SenderProfileEntryResponse;
+        }
+        posts.map((post) => (post.ProfileEntryResponse = res.ReceiverProfileEntryResponse));
+
+        let lastDiamondLevel = this.lastDiamondLevelOnPage[page - 1];
+        console.log(lastDiamondLevel);
+        for (let ii = 0; ii < posts.length; ii++) {
+          posts[ii].ProfileEntryResponse = this.receiverProfileEntryResponse;
+          if (posts[ii].DiamondsFromSender != lastDiamondLevel) {
+            posts[ii].ShowDiamondDivider = true;
```

Comment:
we shouldn't be editing the PostEntryResponse to include rendering data. it should only contain the response from the server.

make a new type called something like DiamondsPost that contains a PostEntryResponse, ShowDiamondDivider, and DiamondsFromSender please.

Reply:
core's PostEntryResponse struct includes `DiamondsFromSender`, but I'll create a new type for DiamondPost that is a post and the show diamond divider attribute